apps/olat-api/src/index.ts

```diff
@@ -0,0 +1,488 @@
+let activityKeysGeneral: string[] = []
+type ActivityTypeSubselection = (typeof activityKeysSubselection)[number]
+
+async function apiKeyMiddleware(req: Request, res: Response, next: () => void) {
+  const apiKey = req.headers['x-api-key']
+  if (!apiKey || typeof apiKey !== 'string') {
+    return res.status(StatusCode.BAD_REQUEST).json({ error: 'Missing API key' })
+  }
+  if (apiKey !== API_KEY) {
+    return res
+      .status(StatusCode.UNAUTHORIZED)
+      .json({ error: 'Invalid API key' })
+  }
+  next()
+}
+
+const limiter = rateLimit({
+  windowMs: 60 * 1000, // 1 minute
+  max: 100,
+  standardHeaders: true, // RateLimit-* headers
+  legacyHeaders: false, // Disable `X-RateLimit-*` headers
+  message: {
+    error: 'Too many requests, please try again later.',
+  },
+})
+
+app.use(limiter)
+app.use('/api', apiKeyMiddleware)
+app.use(express.json())
+
+app.get('/health', (req: Request, res: Response) => {
+  res.status(StatusCode.SUCCESS).json({
+    status: 'OK',
+    timestamp: new Date().toISOString(),
+  })
+})
+
+app.get('/', (req: Request, res: Response) => {
+  res.json({ message: 'OLAT API server' })
+})
+
+// Serve OpenAPI specification
+app.get('/openapi.yaml', async (req: Request, res: Response) => {
+  try {
+    const __filename = fileURLToPath(import.meta.url)
+    const __dirname = path.dirname(__filename)
+    const specPath = path.join(__dirname, '../static/openapi.yaml')
+    const yamlContent = await fs.readFile(specPath, 'utf-8')
+    res.set('Content-Type', 'application/yaml')
+    res.send(yamlContent)
+  } catch (error) {
+    console.error('Error serving OpenAPI specification:', error)
+    res.status(StatusCode.INTERNAL_SERVER_ERROR).json({
+      error: 'Failed to load OpenAPI specification',
+    })
+  }
+})
+
+// Serve Scalar API documentation
+app.use(
+  '/api-docs',
+  apiReference({
+    url: '/openapi.yaml',
+    theme: 'default',
+  })
+)
+
+function checkHeader(req: Request): StatusCode {
+  const contentType = req.headers['content-type']
+  if (!contentType || typeof contentType !== 'string') {
+    return StatusCode.BAD_REQUEST
+  }
+  if (contentType !== 'application/json') {
+    return StatusCode.UNSUPPORTED_MEDIA_TYPE
+  }
+
+  return StatusCode.SUCCESS
+}
+
+async function getCourses(
+  provider: string,
+  providerAccountId: string
+): Promise<any[] | null> {
+  const account = await prisma.account.findUnique({
+    where: {
+      provider_providerAccountId: {
+        provider: provider,
+        providerAccountId: providerAccountId,
+      },
+    },
+    include: {
+      user: {
+        include: {
+          courses: true,
+        },
+      },
+    },
+  })
+
+  const courses = account?.user?.courses ?? []
+  if (courses.length === 0) {
+    return []
+  }
+  const coursesDetails = courses.map((course) => ({
+    id: course.id,
+    title: course.name,
+  }))
+
+  return coursesDetails
+}
+
+app.get('/api/configuration/courses', (req: Request, res: Response) => {
```

Comment:
it is crucial that we move any endpoint that receives parameters to a POST request. this is important for security and follows best practice. so getting the list of activity types is clearly still a GET request, but submitting a user id and getting the users courses needs to be POST, otherwise the user id could be exposed.

see e.g. https://stackoverflow.com/questions/26671599/are-security-concerns-sending-a-password-using-a-get-request-over-https-valid

Reply:
using POST would probably require the addition of body-parser to parse JSON contents of post bodies